torchtext/modules/multiheadattention.py

```diff
@@ -0,0 +1,229 @@
+    def __init__(self, in_proj_tuple, attention_layer, out_proj):
+        r""" A multi-head attention container
+
+        Args:
+            in_proj_tuple: A tuple of multi-head in-projection layers
+            attention_layer: The attention layer.
+            out_proj: The multi-head out-projection layer
+
+        Examples::
+            >>> embed_dim, num_heads, bsz = 10, 5, 64
+            >>> MHA = MultiheadAttentionContainer((MultiheadInProject(embed_dim, num_heads),
+                                                   MultiheadInProject(embed_dim, num_heads),
+                                                   MultiheadInProject(embed_dim, num_heads)),
+                                                   ScaledDotProduct(),
+                                                   MultiheadOutProject(embed_dim // num_heads, num_heads))
+            >>> query = torch.rand((21, bsz, embed_dim))
+            >>> key = value = torch.rand((16, bsz, embed_dim))
+            >>> attn_output, attn_weights = MHA(query, key, value)
+            >>> print(attn_output.shape)
+            >>> torch.Size([21, 64, 10])
+        """
+        super(MultiheadAttentionContainer, self).__init__()
+        self.query_in_proj = in_proj_tuple[0]
+        self.key_in_proj = in_proj_tuple[1]
+        self.value_in_proj = in_proj_tuple[2]
+        self.attention_layer = attention_layer
+        self.out_proj = out_proj
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor,
+                attn_mask: Optional[torch.Tensor] = None,
+                bias_k: Optional[torch.Tensor] = None,
+                bias_v: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+
+        Args:
+            query, key, value (Tensor): map a query and a set of key-value pairs to an output.
+                See "Attention Is All You Need" for more details.
+            attn_mask (BoolTensor, optional): 3D mask that prevents attention to certain positions.
+            bias_k and bias_v:bias (Tensor, optional): one more key and value sequence to be added at
+                sequence dim (dim=-3). Those are used for incremental decoding.
+
+        Shape:
+            - Inputs:
+            - query: :math:`(L, N, E)`
+            - key: :math:`(S, N, E)`
+            - value: :math:`(S, N, E)`
+            - attn_mask: :math:`(N * H, L, S)`, positions with ``True`` are not allowed to attend
+                while ``False`` values will be unchanged.
+            - bias_k and bias_v:bias: :math:`(1, N * H, E / H)`
+
+            - Outputs:
+            - attn_output: :math:`(L, N, E)`
+            - attn_output_weights: :math:`(N*num_heads, L, S)`
+
+            where where L is the target length, S is the sequence length, H is the number of attention heads,
+                N is the batch size, and E is the embedding dimension.
+        """
+        q = self.query_in_proj(query)
```

Comment:
This container doesn't do a lot of work. It requires the users to do all the reshaping and concatenation which was flagged as the hard part